Algorithms/DayOfYear.java

```diff
@@ -0,0 +1,96 @@
+class DayOfYear{
+	public static void main(String[] args) {
+		for (int i = 1980; i <= 2000; i++) {
+			// System.out.println(i + ", " + i%7);
+			// System.out.println(i + ", " + isLeapYear(i));
+		}
+		// System.out.println("1900, " + isLeapYear(1900));
+		// System.out.println("2100, " + isLeapYear(2100));
+		// System.out.println("2200, " + isLeapYear(2200));
+		// System.out.println("2300, " + isLeapYear(2300));
+		// System.out.println("2400, " + isLeapYear(2400));
+
+		// System.out.println(DayOfWeek.values()[0].longName);
+
+		System.out.println("1900, " + calcFirstDayOfYear(1900));
+		System.out.println("1901, " + calcFirstDayOfYear(1901));
+		System.out.println("2005, " + calcFirstDayOfYear(2005));
+		System.out.println("2200, " + calcFirstDayOfYear(2200));
+		// System.out.println(101%2==0);
+
+		//GOAL: calculate day of week for 1/1/y = year
+		//x=?
+		//where x is day of week
+		//1.) calculate initial offset
+		//2.) isLeapYear
+		//3.) number of days from start of the year until x
+		//4.) x % 7
+
+		//after doing the math to find out how many leap years there will be run a check on 
+		//each of them to confirm they are in fact leap years
+
+
+
+	}
+
+	//calculate first day of year
+	//calcFirstDayOfYear()
+	public static String calcFirstDayOfYear(int year) {
+		int leapYearCount;
+		int payload;
+
+		payload = year - 1900; //We consider 1900 the first year ever
+		leapYearCount = payload / 4; //Calculate number of leap years
+		
+		//validate leapYearCount; 25 leap years puts us at the year 2000
+		if (leapYearCount > 25) {
+			int validateLeapYearCount = 24; //reduced by one since we are checking 2000 at start
+
+			for (int i = 2000; i <= year; i++) {
+				//check to see if divisable by 4 before checking isLeapYear, 
+				if (i % 4 == 0) {
+					if (isLeapYear(i)) {
+						validateLeapYearCount++;
+					}
+				}
+			}
+
+			if (leapYearCount != validateLeapYearCount) {
+				leapYearCount = validateLeapYearCount;
+			}
+		}
```

Comment:
I was thinking about it and there is also a mathematical way to calculate this instead of looping for it. The loop itself is the more inefficient since as of 2019 we are only able to go in one direction in time and there will be more years ahead than there had been since 1900. 

* Assume `year = 2507`
* Assume `payload = year - 1900 = 607`
* Assume `leapYearCount = (payload - 1) / 4 = (607 - 1) / 4 = 151`

In this case `leapYearCount` is the number of all leap years between 1900 and 2506 including the following tricky cases:

| Year | Should be leap |
|---|---|
|`1900`|yes|
|`2000`|yes|
|`2100`|**no**|
|`2200`|**no**|
|`2300`|**no**|
|`2400`|yes|
|`2500`|**no**|

So we see that we have 4 more leap years than what they really were. To determine those we can do the following:
```java
// if the payload is greater than 200 (we're asking for at least 2100)
if(payload >= 200) { 

     // payload since 2000 without the intended year
    payload2k = payload - 100 - 1; // 607 - 100 - 1 = 506 

    // we get the century years (which we know already are divisible by 4 )
    leapCenturyYears = payload2k / 100; // 506 / 100 = 5

    // we get the multiples of four-hundred (which we know shouldn't be counted as leap) 
    leap4CenturyYears = payload2k / 400; // 506 / 400 = 1

    // we do an intermediate calculation
    invalidLeapYears = leapCenturyYears - leap4CenturyYears;  // 5 - 1 = 4

    // then we subtract those from the valid number of leap years
    leapYearCount -= invalidLeapYears // 151 - 4 = 147
}
```
And with that:
```java
payload += leapYearCount + 1 // 607 + 147 + 1 = 755
payload %= 7 // 755 % 7 = 6 -> Saturday as expected
```

Reply:
Added this into the code, will need to do on paper for it to make 100% sense to me. Will be in next push.